ksqldb-common/src/main/java/io/confluent/ksql/reactive/BaseSubscriber.java

```diff
@@ -147,7 +147,7 @@ protected final void checkContext() {
     VertxUtils.checkContext(context);
   }
 
-  private void runOnRightContext(final Runnable runnable) {
+  protected void runOnRightContext(final Runnable runnable) {
```

Comment:
I don't think we should expose this method. Let's just update the problematic line in `PollableSubscriber#poll()` to call `context.runOnContext(v -> checkRequestTokens());` instead of `checkRequestTokens()`. 

The purpose of having strict context checks is to ensure that implementers of BaseSubscriber are cognizant of which methods are being run where. When additional, user-facing methods outside the public methods of BaseSubscriber are exposed, it's preferable to have the implementer intentionally use the correct context, rather than throwing a blanket of "use this context if needed" at the method. If you're curious, there's precedent here: https://github.com/confluentinc/ksql/blob/ed2a1860e9c0b85388b41a2f8049d328cc2aba1d/ksqldb-rest-app/src/main/java/io/confluent/ksql/api/impl/InsertsSubscriber.java#L232

Reply:
> I don't think we should expose this method. Let's just update the problematic line in PollableSubscriber#poll() to call context.runOnContext(v -> checkRequestTokens()); instead of checkRequestTokens().

Ok, sounds reasonable.  Rather than make token volatile, which you'd have to do if it was updated in the context, I just changed this within `checkRequestTokens` to:
```
context.runOnContext(v -> makeRequest(REQUEST_BATCH_SIZE));
```
 
This fits the invariant you describe about makeRequest being called on context as it is everywhere else.  Is that reasonable?

Reply:
> Or do all non-public methods generally get called on the context?

Are you asking whether the entire `poll()` call should be run on the context? If so, the answer is no. Blocking operations are not meant to be performed on event loops.

> Rather than use synchonization with token and use an AtomicInteger, which you'd have to do if it was updated in the context, I just changed this within checkRequestTokens to: `context.runOnContext(v -> makeRequest(REQUEST_BATCH_SIZE));`

My only potential concern with this is that because `checkRequestTokens()` is called from `afterSubscribe()` which is already wrapped in a call to `context.runOnContext(...);`, we now have two layers of wrapping in this case. I think this is fine but it'd be good to double-check. (I hope we have automated testing that would fail if this was disallowed/blocking!)

Reply:
> Are you asking whether the entire poll() call should be run on the context? If so, the answer is no. Blocking operations are not meant to be performed on event loops.

I definitely get that that's not allowed on the context.  Poll is a public method, so I wasn't meaning to include that.  I was more generally asking about the convention about when to ensure code is running on the context vs potentially some other thread.  "Do the onramps to the context only happen at a few well defined public API functions in these publishers and subscribers?" was what I meant.
 
> My only potential concern with this is that because checkRequestTokens() is called from afterSubscribe() which is already wrapped in a call to context.runOnContext(...);, we now have two layers of wrapping in this case. I think this is fine but it'd be good to double-check. (I hope we have automated testing that would fail if this was disallowed/blocking!)

Ah, you're right.  Didn't realize that about `afterSubscribe`.  Ok, tokens was already being accessed by multiple threads, so I just made it an AtomicInteger and took your original suggestion.

Reply:
> I was more generally asking about the convention about when to ensure code is running on the context vs potentially some other thread. "Do the onramps to the context only happen at a few well defined public API functions in these publishers and subscribers?" was what I meant.

In general or in how the ksql code in particular is setup? I don't know the general answer but the ksql setup makes sense to me: BaseSubscriber ensures all the subscriber methods are run on the correct context, since otherwise all subscriber implementations would have to do that separately. For all other methods (not included in BaseSubscriber), the subscriber implementation must ensure the relevant calls are run on the correct context, while minimizing the code that is run on the context to what must be run on the context, in order to not consume event loop resources unnecessarily. 

Not sure if I actually answered your question. LMK if I'm still misunderstanding.

Reply:
I think that answers my question.  I was kind of wanting to know if in our codebase we had some kind of philosophy since reasoning at a glance about whether some method in one of these classes is run on or off the context is hard. In this case, it seems like `BaseSubscriber` handles most of that, which makes it easier.  Looking through the code, `PollableSubscriber` seems like a bit of an exception since it provides another public method that we know is run off the context, whereas all of these other uses are all async by nature.